src/AwsClient.php

```diff
@@ -403,48 +403,59 @@ private function addEndpointDiscoveryMiddleware($config, $args)
         }
     }
 
-    private function addSignatureMiddleware()
+    private function addSignatureMiddleware(array $args)
     {
         $api = $this->getApi();
         $provider = $this->signatureProvider;
         $version = $this->config['signature_version'];
         $name = $this->config['signing_name'];
         $region = $this->config['signing_region'];
 
+        if (isset($args['signature_version'])
+         || isset($this->config['configured_signature_version'])
+        ) {
+            $configuredSignatureVersion = true;
+        } else {
+            $configuredSignatureVersion = false;
+        }
+
```

Comment:
Why do we need a flag `configured_signature_version` in order to give preference to the signature_version coming as config?; and even, how is the `signature_version` from the arguments `$args` ever considered?.  Could we maybe populate `$version` from either `$args['signature_version']` or `$this->config['signature_version']`?, depending on which one should be considered first, and later if `$version` is not empty then just ignore any logic that would populate `$version` again. Because, I feel it is not clear what the order of preference is, when selecting which `signature_version` to use.

Reply:
Discussed with Sean in a call regarding this. Basically, `$this->config['configured_signature_version']` will have either the default `signature_version` or a custom provided signature version, and when a custom signature version is provided then it must take priority, and therefore a flag is needed to know whether $this->config['configured_signature_version'] is the default which can be overridden, or if is custom provided one.